src/Components/Web/src/Virtualization/VirtualizeDeferred.cs

```diff
@@ -0,0 +1,138 @@
+        /// </summary>
+        [Parameter]
+        public RenderFragment<TItem> Item { get; set; } = default!;
+
+        /// <summary>
+        /// Gets or sets the template for items that have not yet been loaded in memory.
+        /// </summary>
+        [Parameter]
+        public RenderFragment<int> Placeholder { get; set; } = default!;
+
+        /// <summary>
+        /// Gets or sets the function providing items to the list.
+        /// </summary>
+        [Parameter]
+        public ItemsProviderDelegate<TItem> ItemsProvider { get; set; } = default!;
+
+        /// <inheritdoc />
+        protected override int ItemCount => _itemCount;
+
+        /// <inheritdoc />
+        protected override void OnParametersSet()
+        {
+            if (Item == null)
+            {
+                throw new InvalidOperationException($"{GetType()} requires the template parameter '{nameof(Item)}' to be specified.");
+            }
+
+            if (Placeholder == null)
+            {
+                throw new InvalidOperationException($"{GetType()} requires the template parameter '{nameof(Placeholder)}' to be specified.");
+            }
+
+            if (ItemsProvider == null)
+            {
+                throw new InvalidOperationException($"{GetType()} requires a non-null parameter '{nameof(ItemsProvider)}'.");
+            }
+        }
+
+        /// <inheritdoc />
+        protected override async Task OnAfterRenderAsync(bool firstRender)
+        {
+            await base.OnAfterRenderAsync(firstRender);
+
+            if (_fetchState == FetchState.Waiting)
+            {
+                Debug.Assert(_fetchTask != null);
+
+                _fetchState = FetchState.Busy;
+
+                var result = await _fetchTask;
+
+                _itemCount = result.TotalItemCount;
+
+                foreach (var item in result.Items)
+                {
+                    _loadedItems.Enqueue(item);
+                }
+
+                _fetchTcs = null;
+                _fetchState = FetchState.Idle;
+
+                StateHasChanged();
+            }
+        }
+
+        /// <inheritdoc />
+        protected override void RenderItems(RenderTreeBuilder builder)
+        {
+            var itemsRendered = 0;
+
+            foreach (var item in _loadedItems.Skip(ItemsBefore).Take(ItemsVisible))
+            {
+                Item(item)(builder);
+                itemsRendered++;
+            }
+
+            for (; itemsRendered < ItemsVisible; itemsRendered++)
+            {
+                Placeholder(ItemsBefore + itemsRendered)(builder);
+            }
+
+            var itemsToFetch = ItemsBefore + VisibleItemCapacity - _loadedItems.Count;
+
+            if (itemsToFetch > 0 && _fetchState == FetchState.Idle)
+            {
+                _fetchState = FetchState.Waiting;
+                _fetchTcs = new CancellationTokenSource();
+                _fetchTask = ItemsProvider(_loadedItems.Count, itemsToFetch, _fetchTcs.Token);
```

Comment:
It seems really odd that we trigger the data loading during rendering instead of during the JS interop call that triggers this flow.

Is it possible this causes more repetition of rendering than would otherwise be needed? For example it seems that even if `ItemsProvider` returns a task that has already completed synchronously, we wouldn't actually render those items during this render cycle - we'd do a whole round-trip with the client before the `OnAfterRenderAsync` which would then start another render cycle to render those items.

The traditional sync-or-async rendering flow we use in other places looks like:

 * Something happens to trigger an update (e.g., a DOM event, or in your case, a JS interop call saying that a spacer is becoming visible)
 * We invoke the developer's callback that returns a `Task` (in your case `Task<ItemsProviderResult>`), but we don't await it yet
 * We then render synchronously
 * Then, if the returned task is not yet completed, we await it, followed by rendering again to reflect its async completion

If it's possible to follow that same flow here, it might avoid some unnecessary rendering as well as being more familiar for customers.

Again, if I'm missing an important detail about why the data loading can't be started until we're inside rendering, please let me know :)

Reply:
The main reason I handled the data loading inside rendering was because when relying on the spacer callbacks, there are some cases where more items should be fetched, but aren't, because the spacers did not change size. For instance, suppose we start of with an empty virtualized list tall enough to displaying 5 items at once, but the initial fetch only returns 3. The spacers did not change size because there is no need to simulate scrolling, so the spacer callbacks didn't get invoked, and the user's inability to scroll means no more items could ever be fetched (unless they could shrink the container to create a scrollbar). `<Virtualize>` should be able to recognize that it still needs more items to fill the visible part of the container. I've also found other miscellaneous ways to trick `<Virtualize>` into not fetching items when it should by relying on spacer callbacks.

I could certainly look into a workaround for the cases that behave weirdly, but I do have one follow-up question. One point you brought up was that initiating data fetches during rendering could cause unnecessary renders if the data was fetched synchronously. Couldn't the rendering flow you've described above be compatible with initiating fetches during rendering? It seems we could find out if a fetch is needed at the beginning of `BuildRenderTree`, and if the data was fetched synchronously, the results would be available for the current render. If not, the rest of the flow is identical.